### DISCUSSION TITLE
Support Model Context Protocol (MCP)

### DISCUSSION THREAD

**Discussion by kpavlov** [langchain4j/langchain4j#2204] [Ideas]
## Idea

> The Model Context Protocol (MCP) is built on a flexible, extensible architecture that enables seamless communication between LLM applications and integrations. This document covers the core architectural components and concepts.

1. Support Model Context Protocol (MCP) for integrating with external Tools.
LC4j may use MCP server to build a ToolSpecification and also run Tool requests against MCP server. 
- https://modelcontextprotocol.io/docs/concepts/tools

2. Support Model Context Protocol (MCP) for to retrieve external with Prompts (related to https://github.com/kpavlov/langchain4j/pull/11)
https://modelcontextprotocol.io/docs/concepts/prompts

## Links
- https://github.com/modelcontextprotocol
- https://modelcontextprotocol.io/docs/concepts/architecture
- https://www.anthropic.com/news/model-context-protocol

**Comment by kpavlov** [comment]
Someone is very fast; https://spring.io/blog/2024/12/11/spring-ai-mcp-announcement

**Reply by geoand** [reply]
Fast in writing blog posts for sure :)

**Comment by agoncal** [comment]
It has been added to the official MCP repo the 14th of Feb: https://modelcontextprotocol.io/development/updates

If you look at the structure of the repo (https://github.com/modelcontextprotocol/java-sdk), there is an `mcp` and an `mcp-spring` module. Wonder if it would make sense to have a `mcp-langchain4j` module and how easy it would be to implement

**Reply by smileyboy2019** [reply]
Yes, we still need to push it forward as soon as possible

**Comment by smileyboy2019** [comment]
In the case, only the client-side code of MCP was seen. Is there any server-side code that can be viewed

**Reply by geoand** [reply]
See https://quarkus.io/blog/mcp-server/
